Not compatible with Gradle 7.0

Make compatible with Gradle 7.0

Seeing:

```
FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task 'lib:spoonDebugAndroidTest' (type 'SpoonTask').
  - Type 'com.jaredsburrows.spoon.SpoonTask' property 'applicationApk' is missing an input or output annotation.
    
    Reason: A property without annotation isn't considered during up-to-date checking.
    
    Possible solutions:
      1. Add an input or output annotation.
      2. Mark it as @Internal.
    
    Please refer to https://docs.gradle.org/7.0.2/userguide/validation_problems.html#missing_annotation for more details about this problem.
```

As `applicationApk` is already marked with `@Internal` - maybe it needs to be marked with `@get:Internal` due to it being a `lateinit`?

https://github.com/jaredsburrows/gradle-spoon-plugin/blob/b9f5afa843e740d5737c3964018d780aad73d54f/src/main/kotlin/com/jaredsburrows/spoon/SpoonTask.kt#L20-L26

There's also issues on top of this due to the use of ddmlib which has changed certainly with build tools 30.

```
Caused by: java.lang.NoSuchMethodError: com.android.ddmlib.logcat.LogCatMessage.getTimestamp()Lcom/android/ddmlib/logcat/LogCatTimestamp;
        at com.squareup.spoon.html.HtmlLog$LogEntry.from(HtmlLog.java:68)
        at com.squareup.spoon.html.HtmlLog.from(HtmlLog.java:36)
        at com.squareup.spoon.html.HtmlRenderer.generateLogHtml(HtmlRenderer.java:145)
        at com.squareup.spoon.html.HtmlRenderer.render(HtmlRenderer.java:63)
```

This is an issue with spoon itself, which it seems isn't getting very much in the way of updates.  They haven't put out a release in a long time, and has seen 2 commits in 3 years.

--- a/gradle-spoon-plugin/src/test/groovy/com/jaredsburrows/spoon/SpoonPluginVersionsSpec.groovy
+++ b/gradle-spoon-plugin/src/test/groovy/com/jaredsburrows/spoon/SpoonPluginVersionsSpec.groovy
@@ -1,0 +1,412 @@
+package com.jaredsburrows.spoon
+
+import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
+
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import spock.lang.Specification
+import spock.lang.Unroll
+
+final class SpoonPluginVersionsSpec extends Specification {
+  @Rule public final TemporaryFolder testProjectDir = new TemporaryFolder()
+  private int compileSdkVersion = 28
+  private List<File> pluginClasspath
+  private String classpathString
+  private File buildFile
+
+  def 'setup'() {
+    def pluginClasspathResource = getClass().classLoader.getResource('plugin-classpath.txt')
+    if (pluginClasspathResource == null) {
+      throw new IllegalStateException(
+        'Did not find plugin classpath resource, run `testClasses` build task.')
+    }
+
+    pluginClasspath = pluginClasspathResource.readLines().collect { new File(it) }
+    classpathString = pluginClasspath
+      .collect { it.absolutePath.replace('\\', '\\\\') } // escape backslashes in Windows paths
+      .collect { "'$it'" }
+      .join(", ")
+    buildFile = testProjectDir.newFile('build.gradle')
+  }
+
+  @Unroll def 'AGP version 3.6+, gradle: #gradleVersion and AGP: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 5.6.4+, 3.6.0-3.6.4
+    [gradleVersion, agpVersion] << [
+      [
+        '7.0.2',
+        '7.1.1',
+        '7.2',
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '3.6.4',
+      ]
+    ].combinations()
+  }
+
+  @Unroll def 'AGP version 4.0+, gradle: #gradleVersion and AGP: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 6.1.1+, 4.0.0+
+    [gradleVersion, agpVersion] << [
+      [
+        '7.0.2',
+        '7.1.1',
+        '7.2',
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '4.0.2',
+      ]
+    ].combinations()
+  }
+
+  @Unroll def 'agp version 4.1+, gradle: #gradleVersion and agp: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 6.5+, 4.1.0+
+    [gradleVersion, agpVersion] << [
+      [
+        '7.0.2',
+        '7.1.1',
+        '7.2',
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '4.1.3',
+      ]
+    ].combinations()
+  }
+
+  @Unroll def 'agp version 4.2+, gradle: #gradleVersion and agp: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 6.7.1+, 4.2.0+
+    [gradleVersion, agpVersion] << [
+      [
+        '7.0.2',
+        '7.1.1',
+        '7.2',
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '4.2.2',
+      ]
+    ].combinations()
+  }
+
+  @Unroll def 'agp version 7.0+, gradle: #gradleVersion and agp: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 7.0+, 7.0
+    [gradleVersion, agpVersion] << [
+      [
+        '7.0.2',
+        '7.1.1',
+        '7.2',
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '7.0.4',
+      ]
+    ].combinations()
+  }
+
+  @Unroll def 'agp version 7.1+, gradle: #gradleVersion and agp: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 7.2+, 7.1
+    [gradleVersion, agpVersion] << [
+      [
+        '7.2',
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '7.1.3',
+      ]
+    ].combinations()
+  }
+
+  @Unroll def 'agp version 7.2+, gradle: #gradleVersion and agp: #agpVersion'() {
+    given:
+    buildFile <<
+      """
+      buildscript {
+        repositories {
+          mavenCentral()
+          google()
+        }
+        dependencies {
+          classpath "com.android.tools.build:gradle:${agpVersion}"
+          classpath files($classpathString)
+        }
+      }
+
+      apply plugin: 'com.android.application'
+      apply plugin: 'com.jaredsburrows.spoon'
+
+      android {
+        compileSdkVersion $compileSdkVersion
+
+        defaultConfig {
+          applicationId 'com.example'
+        }
+      }
+      """
+
+    when:
+    def result = GradleRunner.create()
+      .withGradleVersion(gradleVersion)
+      .withProjectDir(testProjectDir.root)
+      .withArguments('spoonDebugAndroidTest', '-s', '-Ptesting')
+      .build()
+
+    then:
+    result.task(':spoonDebugAndroidTest').outcome == SUCCESS
+
+    where:
+    // https://docs.gradle.org/current/userguide/compatibility.html
+    // https://developer.android.com/studio/releases/gradle-plugin
+    // 7.3+, 7.2
+    [gradleVersion, agpVersion] << [
+      [
+        '7.3.3',
+        '7.4.2',
+      ],
+      [
+        '7.2.0-rc01',
+      ]
+    ].combinations()
+  }
+}

